server/src/test/java/org/elasticsearch/cluster/service/MasterServiceTests.java

```diff
@@ -982,7 +992,12 @@ public void onResponse(ClusterState clusterState) {
 
                         @Override
                         public void onFailure(Exception e) {
-                            throw new AssertionError(e);
+                            assertTrue(expectFailure);
+                            assertThat(e, instanceOf(ElasticsearchException.class));
+                            assertThat(e.getMessage(), equalTo(exceptionMessage));
+                            assertEquals(List.of(testResponseHeaderValue), threadContext.getResponseHeaders().get(testResponseHeaderName));
+                            assertNotNull(publishedState.get());
```

Comment:
It is not entirely clear to me that this could not be the state returned from the `blockMasterTask`? Perhaps we can pollute the cluster state returned from the real executor with a piece of data to check the presence of?

Reply:
The `blockMasterTask` returns an unchanged state so it isn't published. I added some more machinery to verify this in 34e0605cafed1aea283744f2056e7f1af79c730a.